core/src/main/scala/kafka/controller/ReplicaStateMachine.scala

```diff
@@ -357,11 +357,11 @@ class ReplicaStateMachine(controller: KafkaController) extends Logging {
         if (hasStarted.get) {
           ControllerStats.leaderElectionTimer.time {
             try {
+              val curBrokers = currentBrokerList.map(_.toInt).toSet.flatMap(zkUtils.getBrokerInfo)
+              val curBrokerIds = curBrokers.map(_.id)
               val liveOrShuttingDownBrokerIds = controllerContext.liveOrShuttingDownBrokerIds
-              val curBrokerIds = currentBrokerList.map(_.toInt).toSet
               val newBrokerIds = curBrokerIds -- liveOrShuttingDownBrokerIds
               val deadBrokerIds = liveOrShuttingDownBrokerIds -- curBrokerIds
-              val curBrokers = curBrokerIds.flatMap(zkUtils.getBrokerInfo)
               val brokerById = curBrokers.map(broker => broker.id -> broker).toMap
               val newBrokers = newBrokerIds.flatMap(brokerById.get)
```

Comment:
Indeed, it could be replaced by `newBrokerIds.map(brokerById)`.

Another optimization would be to compute `brokerById` only for ids in `newBrokerIds`. It would avoid to compute mapping of 1000 brokers to lookup only few new brokers. wdyt?